### ISSUE TITLE
Relative require paths

### ISSUE THREAD

**Issue by aleclarson** [tarantool/tarantool#3134]
I'm not 100% sure how Luvit wraps each module with its own `require` function, but I really like the idea. Particularly for relative paths' sake.

https://github.com/luvit/luvit/blob/master/deps/require.lua

Is this something Tarantool would consider supporting out-of-the-box?

**Comment by kostja** [comment]
Could you please give a link to the docs or describe the idea in  a greater detail for those who never used luvit?

**Comment by aleclarson** [comment]
After further research, they now use a [package loader][1] instead of creating a dedicated `require` function per module via code-wrapping: https://github.com/luvit/lit/blob/master/luvit-loader.lua

Their `require` system works very similarly to NodeJS:

- Relative paths are supported (eg: `./init` for sibling modules or `../init` to traverse up directories)
- Module names still work (eg: `foo/bar` or `foo.bar`)

There's probably more to it than that, but luvit's documentation is sparse on the subject. Anyway, the relative `require` calls are reason enough to have something like this, in my opinion.

Someone made a [npm][3]-like package manager for luvit, called [lum][4]. Its benefits are listed [here](). Although it's not mature, perhaps there are some things to learn from its design that can be applied to Tarantool's package manager. Easy dependency declaration via `package.lua` and installation of project dependencies into the `./modules` directory.

[1]: http://pgl.yoyo.org/luai/i/package.loaders
[3]: https://en.wikipedia.org/wiki/Npm_(software)
[4]: https://github.com/radare/lum
[5]: https://groups.google.com/d/msg/luvit/MPGM3rKzDrw/5TQw9jC6628J

**Comment by aleclarson** [comment]
Here's a working implementation that you can drop into Tarantool without a fuss.

https://github.com/aleclarson/lula/blob/4b93c4ced4b18db360120af16b8c8c20dab65f63/loader.lua

Would a feature like this belong in Tarantool core, @kostja? Very comfortable for people coming from a NodeJS background.

**Comment by aleclarson** [comment]
This implementation works good in Tarantool: https://github.com/aleclarson/lula/blob/master/loader.lua

It's more comfortable for people coming from a NodeJS background.

```lua
-- Load 'foo.lua' or 'foo/init.lua' sibling
require "./foo"

-- Load 'foo.lua' or 'foo/init.lua' in parent directory
require "../foo"

-- Traverse multiple parents
require "../../foo/bar"
```

To keep the implementation lightweight, only modules within the working directory can use relative paths. The working directory is determined using `os.getenv('PWD')`, but that may be broken for some platforms (tested on OS X).

Would a feature like this belong in Tarantool core, @kostja?

**Comment by rosik** [comment]
Could you please provide an example when such syntax may be useful? Especially `require('../foo')`

**Comment by aleclarson** [comment]
@rosik Packages in NodeJS use relative requires for all internal modules. The same use case exists in Lua (or any module-based language).

**Comment by Totktonada** [comment]
@rosik We use the following hack intensively in tests:

```
-- require in-repo version of graphql/ sources despite current working directory         
package.path = fio.abspath(debug.getinfo(1).source:match("@?(.*/)")                      
    :gsub('/./', '/'):gsub('/+$', '')) .. '/../../?.lua' .. ';' .. package.path   
```

I would prefer to use short alias like `relative_require()`.

**Comment by Totktonada** [comment]
Now @RunsFor works on the similar feature in PR #4193. I think it is more appropriate way to do 'relative require', because allows to leave just module names in an application, but configure a way to find it (in other words, 'file hierarchy' or 'layout') separately.

**Comment by Totktonada** [comment]
Now it is merged into master in 2.2.0-420-g4e27b0e61. I would consider this issue as resolved.

**Comment by Mons** [comment]
We have no further plans to take any action. For local path search `setsearchroot` is intended.
Anyway, require by path is not from lua world.
